compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -122,27 +121,38 @@ struct LoweringContext<'a, 'hir: 'a> {
     is_in_trait_impl: bool,
     is_in_dyn_type: bool,
 
-    /// What to do when we encounter an "anonymous lifetime
-    /// reference". The term "anonymous" is meant to encompass both
-    /// `'_` lifetimes as well as fully elided cases where nothing is
-    /// written at all (e.g., `&T` or `std::cell::Ref<T>`).
-    anonymous_lifetime_mode: AnonymousLifetimeMode,
-
     /// Used to create lifetime definitions for anonymous lifetimes.
     /// When an anonymous lifetime is encountered in a function or impl header and
     /// requires to create a fresh lifetime parameter, it is added
     /// to this list. The results of this list are then added to the list of
     /// lifetime definitions in the corresponding impl or function generics.
-    lifetimes_to_define: Vec<(Span, NodeId)>,
+    lifetimes_to_define: FxIndexMap<NodeId, Span>,
 
     /// If anonymous lifetimes are being collected, this field holds the parent
     /// `LocalDefId` to create the fresh lifetime parameters' `LocalDefId`.
     is_collecting_anonymous_lifetimes: Option<LocalDefId>,
 
     /// Currently in-scope lifetimes defined in impl headers, fn headers, or HRTB.
-    /// We always store a `normalize_to_macros_2_0()` version of the param-name in this
-    /// vector.
-    in_scope_lifetimes: Vec<ParamName>,
+    in_scope_lifetimes: Vec<(ParamName, LocalDefId)>,
+
+    /// Used to handle lifetimes appearing in impl-traits.  When we lower a lifetime,
+    /// it is inserted in the `FxHashMap`, and the resolution is modified so to point
+    /// to the lifetime parameter impl-trait will generate.
+    /// When traversing `for<...>` binders, they are inserted in the `FxHashSet` so
+    /// we know *not* to rebind the introduced lifetimes.
+    captured_lifetimes: Option<(
+        LocalDefId, // parent def_id for new definitions
+        FxHashMap<
+            LocalDefId, // original parameter id
+            (
+                Span,        // Span
+                NodeId,      // synthetized parameter id
+                ParamName,   // parameter name
+                LifetimeRes, // original resolution
+            ),
+        >,
+        FxHashSet<NodeId>, // traversed binders, to ignore
+    )>,
```

Comment:
Is this a temporary measure, i.e. does some later PR remove or change it?
If not then maybe this should be converted into a structure(s) with named fields.